### ISSUE TITLE
Support for Turso

### ISSUE THREAD

**Issue by janpio** [prisma/prisma#19408]
> [Turso](https://turso.tech/) is an edge-hosted, distributed database based on [libSQL](https://libsql.org/), an open-source and open-contribution fork of [SQLite](https://sqlite.org/). It was designed to minimize query latency for applications where queries come from anywhere in the world. In particular, it works well with edge functions provided by cloud platforms such as CloudFlare, Netlify, and Vercel, by putting your data geographically close to the code that accesses it.

https://turso.tech/

**Comment by JohnC0de** [comment]
No updates on this yet? Really?

**Comment by janpio** [comment]
Yes, really. We are working on this, but have nothing to share yet.

**Comment by janpio** [comment]
Hey, we have been working on this for the past few weeks. Soon, we'll have something to share and test! 
If you are interested in helping us, please respond to this super short survey and leave your email address so we can get in touch with instructions: https://pris.ly/survey/driver-adapter-turso Thanks.

**Comment by sankalpmukim** [comment]
![image](https://github.com/prisma/prisma/assets/41541375/1a9c7f43-a0e3-4fe0-83dc-78eb6af7b4f5)
The survey is broken. :)

**Comment by janpio** [comment]
That was hopefully only a 1-time occurrence. It's working for me again, and should for you as well.

**Comment by devjmetivier** [comment]
@janpio Could you comment on how schema migrations would work with Turso? Racking my brain a bit here trying to imagine how that works with a service that has 1 or many databases.

**Comment by janpio** [comment]
For now: You generate the SQL script with `migrate dev`, `migrate diff` or similar Prisma command, then pass it to the Turso CLI shell. Something like this:
```
npx prisma migrate diff --from-empty --to-schema-datamodel prisma/schema.prisma --script > init.sql
turso db shell database-name < init.sql
```

Later, we will want to integrate that a bit better.

(For many databases, we don't have more thoughts yet - that is a [_very_ recent Turso development](https://blog.turso.tech/turso-radically-increases-the-amount-of-databases-available-on-their-starter-and-scaler-plans-10a69ad94055))

**Comment by devjmetivier** [comment]
Yeah my concern was on the side of the _many databases_ case. I see now that it's a more complex problem than I anticipated.

Seems unfortunately the case is that using Prisma (or any ORM for that matter) with multi-tenancy isn't possible without "migrations on the fly".

**Comment by janpio** [comment]
You can totally script running Prisma CLI for all your databases for example, but you are most probably searching for something more convenient and automated - and that we do not have yet.

**Comment by janpio** [comment]
But, I have good news: 

**Support for Prisma and Turso today was launched as Early Access!**
Read more here: https://www.prisma.io/blog/prisma-turso-ea-support-rXGd_Tmy3UXX

**Comment by frontsideair** [comment]
I'm giving this a try and everything worked well so far. To get the best of edge computing, I'm experimenting on moving one route to the edge runtime in Vercel and the build failed, since `@prisma/client` depends on node APIs, which is not a surprise.

My question is, does Turso adapter work with `@prisma/client/edge`, since I didn't see it in the introductory blog post. I would assume `@libsql/client` is doing the heavy lifting and node APIs are only in the module graph and not actually used, and if that's the case there should be an entry point that's compatible with edge workers.

**Comment by janpio** [comment]
It does not yet @frontsideair, but the driver adapter was indeed the first step to get there. The next step is tracked here: https://github.com/prisma/prisma/issues/21394

**Comment by feri-irawan** [comment]
Is Prisma Studio available for Turso now?

**Comment by empflow** [comment]
> You can totally script running Prisma CLI for all your databases for example, but you are most probably searching for something more convenient and automated - and that we do not have yet.

Can you provide an example of such a script since you don't have a better solution just yet? Is it supposed to run as a github action? I don't quite get it

**Comment by janpio** [comment]
Support for Turso is available as an open Early Access since a while:

- Launch blog post: https://www.prisma.io/blog/prisma-turso-ea-support-rXGd_Tmy3UXX
- Documentation: https://www.prisma.io/docs/orm/overview/databases/turso
- Npm package: https://npmjs.com/@prisma/adapter-libsql
- Feedback discussion: https://github.com/prisma/prisma/discussions/21345

If you want to give feedback, please use the [feedback discussion](https://github.com/prisma/prisma/discussions/21345) or open a new bug issue or feature request. Thank you.
